src/VisualStudio/Core/Def/RoslynPackage.cs

```diff
@@ -140,76 +140,94 @@ protected override void OnSaveOptions(string key, Stream stream)
             {
                 _solutionEventMonitor = new SolutionEventMonitor(globalNotificationService);
                 TrackBulkFileOperations(globalNotificationService);
 
                 RegisterEditorFactory(settingsEditorFactory);
 
-                // Proffer in-process service broker services
-                var serviceBrokerContainer = await this.GetServiceAsync<SVsBrokeredServiceContainer, IBrokeredServiceContainer>(this.JoinableTaskFactory).ConfigureAwait(false);
-
-                serviceBrokerContainer.Proffer(
-                    WorkspaceProjectFactoryServiceDescriptor.ServiceDescriptor,
-                    (_, _, _, _) => ValueTaskFactory.FromResult<object?>(new WorkspaceProjectFactoryService(this.ComponentModel.GetService<IWorkspaceProjectContextFactory>())));
-
-                // Must be profferred before any C#/VB projects are loaded and the corresponding UI context activated.
-                serviceBrokerContainer.Proffer(
-                    ManagedHotReloadLanguageServiceDescriptor.Descriptor,
-                    (_, _, _, _) => ValueTaskFactory.FromResult<object?>(new ManagedEditAndContinueLanguageServiceBridge(this.ComponentModel.GetService<EditAndContinueLanguageService>())));
-
                 // Misc workspace has to be up and running by the time our package is usable so that it can track running
                 // doc events and appropriately map files to/from it and other relevant workspaces (like the
                 // metadata-as-source workspace).
                 var miscellaneousFilesWorkspace = this.ComponentModel.GetService<MiscellaneousFilesWorkspace>();
-            });
 
-        return Task.CompletedTask;
+                return Task.CompletedTask;
+            });
     }
 
-    protected override async Task OnAfterPackageLoadedAsync(CancellationToken cancellationToken)
+    protected override void RegisterOnAfterPackageLoadedAsyncWork(PackageLoadTasks afterPackageLoadedTasks)
     {
-        await base.OnAfterPackageLoadedAsync(cancellationToken).ConfigureAwait(false);
+        base.RegisterOnAfterPackageLoadedAsyncWork(afterPackageLoadedTasks);
 
-        // Ensure the options persisters are loaded since we have to fetch options from the shell
-        LoadOptionPersistersAsync(this.ComponentModel, cancellationToken).Forget();
+        afterPackageLoadedTasks.AddTask(isMainThreadTask: false, task: OnAfterPackageLoadedBackgroundThreadAsync);
+        afterPackageLoadedTasks.AddTask(isMainThreadTask: true, task: OnAfterPackageLoadedMainThreadAsync);
 
-        await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+        return;
+
+        Task OnAfterPackageLoadedBackgroundThreadAsync(PackageLoadTasks afterPackageLoadedTasks, CancellationToken cancellationToken)
+        {
+            // Ensure the options persisters are loaded since we have to fetch options from the shell
+            LoadOptionPersistersAsync(this.ComponentModel, cancellationToken).Forget();
```

Comment:
This is still hopping back to the main thread for the PackageSettingsPersister, right? Will that cause any problems? I don't imagine so since you're still doing a JTF friendly await but I don't imagine we risk deadlock, right?

Reply:
Can you point me to the main thread switch you're mentioning.

Reply:
Oh never mind I was wrong it doesn't exist.